src/publication_cache.rs

```diff
@@ -0,0 +1,196 @@
+//
+// Copyright (c) 2023 ZettaScale Technology.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh team, <zenoh@zettascale.tech>
+//
+
+use std::ops::Deref;
+
+use zenoh_ext::SessionExt;
+use zenoh_util::core::zresult::ErrNo;
+use zenoh_util::core::SyncResolve;
+
+use crate::{
+    impl_guarded_transmute, z_keyexpr_t, z_locality_t, z_session_t, GuardedTransmute,
+    UninitializedKeyExprError,
+};
+
+/// Options passed to the :c:func:`ze_declare_publication_cache` function.
+///
+/// Members:
+///     queryable_prefix: the prefix used for queryable
+///     queryable_origin: the restriction for the matching queries that will be receive by this
+///                       publication cache
+///     history: the the history size
+///     resources_limit: the limit number of cached resources
+#[repr(C)]
+pub struct ze_publication_cache_options_t {
+    pub queryable_prefix: z_keyexpr_t,
+    pub queryable_origin: z_locality_t,
+    pub history: usize,
+    pub resources_limit: usize,
+}
+
+/// Constructs the default value for :c:type:`ze_publication_cache_options_t`.
+#[no_mangle]
+pub extern "C" fn ze_publication_cache_options_default() -> ze_publication_cache_options_t {
+    ze_publication_cache_options_t {
+        queryable_prefix: z_keyexpr_t::null(),
+        queryable_origin: z_locality_t::ANY,
+        history: 1,
+        resources_limit: 0,
+    }
+}
+
+type PublicationCache = Option<Box<zenoh_ext::PublicationCache<'static>>>;
+
+/// An owned zenoh publication_cache.
+///
+/// Like most `z_owned_X_t` types, you may obtain an instance of `z_X_t` by loaning it using `z_X_loan(&val)`.  
+/// The `z_loan(val)` macro, available if your compiler supports C11's `_Generic`, is equivalent to writing `z_X_loan(&val)`.  
+///
+/// Like all `z_owned_X_t`, an instance will be destroyed by any function which takes a mutable pointer to said instance, as this implies the instance's inners were moved.  
+/// To make this fact more obvious when reading your code, consider using `z_move(val)` instead of `&val` as the argument.  
+/// After a move, `val` will still exist, but will no longer be valid. The destructors are double-drop-safe, but other functions will still trust that your `val` is valid.  
+///
+/// To check if `val` is still valid, you may use `z_X_check(&val)` or `z_check(val)` if your compiler supports `_Generic`, which will return `true` if `val` is valid.
+#[cfg(not(target_arch = "arm"))]
+#[repr(C, align(8))]
+pub struct ze_owned_publication_cache_t([u64; 1]);
```

Comment:
Since you have a pointer in there, you can just use `usize` which is guaranteed to be pointer sized and aligned.